p2panda-rs/src/document/document.rs

```diff
@@ -0,0 +1,474 @@
+    #[allow(dead_code)]
+    permissions: Option<Vec<Author>>,
+    /// A map of all operations contained within this document. This may even include operations by unauthorized authors.
+    operations: BTreeMap<String, OperationWithMeta>,
+    /// A causal graph representation of this documents operations, identified by their hash, which can be topologically sorted.
+    graph: IncrementalTopo<String>,
+}
+
+impl Document {
+    /// The hash id of this document.
+    pub fn id(&self) -> Hash {
+        self.id.clone()
+    }
+
+    /// The hash id of this documents schema.
+    pub fn schema(&self) -> Hash {
+        self.schema.schema_hash()
+    }
+
+    /// The author of this document.
+    pub fn author(&self) -> Author {
+        self.author.clone()
+    }
+
+    /// Returns a map of all operations in this document.
+    pub fn operations(&self) -> BTreeMap<String, OperationWithMeta> {
+        self.operations.clone()
+    }
+
+    /// Get the create operation for this document. We unwrap and panic if the value is None
+    /// as all documents should contain at least a create message. This was validated when building
+    /// with DocumentBuilder.
+    fn get_create_operation(&self) -> OperationWithMeta {
+        self.get_operation(self.id().as_str())
+            .expect("There should be a CREATE operation")
+    }
+
+    /// Get an operation from this document by its id. Returns an error if operation
+    /// is not found.
+    fn get_operation(&self, id: &str) -> Result<OperationWithMeta, DocumentError> {
+        match self.operations.get(id) {
+            Some(operation) => Ok(operation.to_owned()),
+            None => Err(DocumentError::OperationNotFound),
+        }
+    }
+
+    /// Returns a result containing an iterable collection of topologically sorted
+    /// nodes in this graph, _except_ the root, identified by their id.
+    fn sort(&self) -> Result<incremental_topo::Descendants<String>, DocumentError> {
+        match self.graph.descendants(self.id().as_str()) {
+            Ok(d) => Ok(d),
+            Err(_) => Err(DocumentError::IncrementalTopoError),
+        }
+    }
+
+    /// Sort the graph topologically, then reduce the linearised operations into a single
+    /// `Instance`.
+    pub fn resolve(&self) -> Result<Instance, DocumentError> {
+        let mut instance = Instance::try_from(self.get_create_operation())?;
+
+        self.sort()?
+            .try_for_each(|id| match instance.apply_update(self.get_operation(id)?) {
+                Ok(_) => Ok(()),
+                Err(e) => Err(DocumentError::InstanceError(e)),
+            })?;
+        Ok(instance)
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl Validate for Document {
+    type Error = DocumentError;
+
+    fn validate(&self) -> Result<(), Self::Error> {
+        // NB. This validation is quite excessive as it's normally not possible to get to this
+        // point while having broken some of these basic data restraints.
```

Comment:
Interesting comment. What are the "basic data restraints". Is it worth putting this into an issue?

Reply:
I was thinking of things like a DELETE operation not containing fields, an UPDATE and CREATE containing previous_operations, things like that. Right now we're validating these criteria when building the document, but I'm not sure if it's necessary.

What issue would you want to raise for this? Something like "When does validation occur in the materialisation pipeline"?

Reply:
It feels a bit like schema validation is something that should be done by operations themselves, doesn't it? They contain both the fields and the schema that's supposedly being used. And we eventually won't handle operations whose schema definition we don't have. Maybe `OperationWithMeta` should contain the operation's schema definition and check it in its validation and `Operation` doesn't?

Reply:
I moved one of the checks here to operation already :o